sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/implementation/StorageAccountMsiHandler.java

```diff
@@ -0,0 +1,272 @@
+                        storageAccount.updateParameters.identity().withType(IdentityType.SYSTEM_ASSIGNED);
+                    } else if (currentIdentity.type().equals(IdentityType.USER_ASSIGNED)) {
+                        storageAccount.updateParameters.identity().withType(IdentityType.NONE);
+                    }
+                    // and set identities property in the payload model to null so that it won't be sent
+                    storageAccount.updateParameters.identity().withUserAssignedIdentities(null);
+                    return true;
+                } else {
+                    // Check user is asking to remove identities though there is no identities currently associated
+                    if (currentIds.isEmpty() && !removeIds.isEmpty() && currentIdentity == null) {
+                        // If so we are in a invalid state but we want to send user input to service and let service
+                        // handle it (ignore or error).
+                        storageAccount.updateParameters.withIdentity(new Identity().withType(IdentityType.NONE).withUserAssignedIdentities(null));
+                        return true;
+                    }
+                }
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Initialize storage account's identity property.
+     *
+     * @param identityType the identity type to set
+     */
+    private void initStorageAccountIdentity(IdentityType identityType) {
+        if (!identityType.equals(IdentityType.USER_ASSIGNED)
+            && !identityType.equals(IdentityType.SYSTEM_ASSIGNED)) {
+            throw logger.logExceptionAsError(new IllegalArgumentException("Invalid argument: " + identityType));
+        }
+        if (storageAccount.isInCreateMode()) {
+            if (storageAccount.createParameters.identity() == null
+                || storageAccount.createParameters.identity().type() == null
+                || storageAccount.createParameters.identity().type().equals(IdentityType.NONE)
+                || storageAccount.createParameters.identity().type().equals(identityType)) {
+                storageAccount.createParameters.withIdentity(new Identity().withType(identityType));
+            } else {
+                storageAccount.createParameters.withIdentity(new Identity().withType(IdentityType.SYSTEM_ASSIGNED_USER_ASSIGNED));
+            }
+        } else {
+            if (storageAccount.innerModel().identity() == null
+                || storageAccount.innerModel().identity().type() == null
+                || storageAccount.innerModel().identity().type().equals(IdentityType.NONE)
+                || storageAccount.innerModel().identity().type().equals(identityType)) {
+                Identity identity = Objects.isNull(storageAccount.innerModel().identity()) ? new Identity().withType(identityType) : storageAccount.innerModel().identity().withType(identityType);
+                storageAccount.updateParameters.withIdentity(identity);
```

Comment:
1. This part is correct to check innerModel (as current Identity configure).

2. However, it then update `updateParameters`.

3. Next time it get called, it would again check innerModel (that does not contain the "update" in (2)).

Reply:
Should we do here similar to `withoutLocalManagedServiceIdentity()`?

Have this above the block
```java
        if (Objects.isNull(storageAccount.updateParameters.identity())) {
            storageAccount.updateParameters.withIdentity(storageAccount.innerModel().identity());
        }
```

And then just check/update `updateParameters`.

@XiaofeiCao Thought?

Reply:
Yeah, I think we should. 
Though current code's behavior is the same, it only has implicit indication that `innerModel()` and `updateParameters` share the same `Identity` instance, which makes the code hard to reason about.